stdlib/atexit.c

```diff
@@ -16,30 +16,111 @@
 #include <stdlib.h>
 #include <string.h>
 
-void (*atexit_funcs[ATEXIT_MAX])(void);
 
+/* The atexit_node structure uses uint32_t for flags that
+ indicate whether to call a function with arguments or not, therefore
+ why ATEXIT_MAX must be 32 as type "fntype" */
+struct atexit_node {
+	void (*destructors[ATEXIT_MAX])(void);
+	void *args[ATEXIT_MAX];
+	uint32_t fntype;
+	struct atexit_node *prev;
+};
 
-void _atexit_call(void)
-{
-	int i;
 
-	for (i = ATEXIT_MAX - 1; i >= 0; i--) {
-		if (atexit_funcs[i] != NULL)
-			atexit_funcs[i]();
-	}
-}
+/* The first node to provide at least 32 function slots */
+static struct atexit_node first_node;
 
 
-#warning FIXME: atexit() uses GCC built-in function defined in Intel Pentium ABI! Works only with -mcpu=i586
+static struct {
+	handle_t lock;
+	struct atexit_node *head;
+} atexit_common = { .head = &first_node };
 
-int atexit(void (*func)(void))
+
+/* Initialise atexit_common structure before main */
+void _atexit_init(void)
+{
+	mutexCreate(&atexit_common.lock);
+}
+
+
+/* Generic function to register destructors */
+static int _atexit_register(int type, void (*fn)(void), void *arg)
 {
+	struct atexit_node *node = atexit_common.head;
 	int i;
 
+	mutexLock(atexit_common.lock);
+	/* Looking for an empty slot */
 	for (i = 0; i < ATEXIT_MAX; i++) {
-		if (__sync_bool_compare_and_swap(&atexit_funcs[i], NULL, func))
+		if (node->destructors[i] == NULL) {
+			if (CXA_ATEXIT_TYPE == type) {
+				node->fntype |= 1u << i;
+				node->args[i] = arg;
+			}
+			node->destructors[i] = fn;
+
+			mutexUnlock(atexit_common.lock);
 			return 0;
+		}
+
+		/* If there is no more functions slots */
+		if (i == ATEXIT_MAX - 1) {
+			/* to allocate new node */
+			node = (struct atexit_node *)calloc(1, sizeof(struct atexit_node));
+			if (node == NULL) {
+				mutexUnlock(atexit_common.lock);
+				return -1;
+			}
+			node->prev = atexit_common.head;
+			atexit_common.head = node;
+
+			/* to start the loop again */
+			i = -1;
+		}
```

Comment:
It doesn't look good: there is no need to check `i` inside loop and restart the loop on new node allocation. I suggest using this scheme:

for (i = 0; i < ATEXIT_MAX; i++) {
  if (have free slot)
    break;
}
if (i == ATEXIT_MAX) {
   alloc  new node
   if (alloc fails)
     return;
  update  head and set i=0
}
assign function to slot i